Update to syntax breakdown plus editorial improvements (#263)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

Within the JSONPath expression, the abstract name `$` is used to refer
to the *root node* of the argument, i.e., to the argument as a whole.

JSONPath expressions use the *bracket notation*, for example:

~~~~
$['store']['book'][0]['title']
~~~~

or the more compact *dot notation*, for example:

~~~~
$.store.book[0].title
~~~~

to build paths that are input to a JSONPath implementation.
A single path may use a combination of bracket and dot notations.

> A wildcard `*` ({{wildcard}}) may be used to select all children of an

Reviewer: wildcards are also available for all descendants.

This should be fixed in the context of making wildcards selectors.

Author: I believe that context is the current PR. The wildcard selector is defined in section 3.5.2.

I'll have to think about how best to address this because a selector acts on a single node and selects zero or more of its children. A selector _per se_ never selects other descendants. But at this stage in the document, we aren't ready to talk about segments and selectors, so we may need to fudge the wording here.

Author: See the rewording in e6ef8e8460b2c5a6f629811f8d8728f790bf5d86.

Reviewer: OK, that works, but could be improved later.  Keeping this open.

object or an array.

An array slice `start:end:step` ({{slice}}) selects a series of
elements from an array, giving a start position, an end position, and
possibly a step value that moves the position from the start to the
end.

Filter expressions `?<boolean expr>` select certain children of an object or array as in

~~~~
$.store.book[?@.price < 10].title
~~~~

> {{tbl-overview}} provides a quick overview of the JSONPath syntax elements.

Reviewer: I believe we should be explicit very early about some of the segments being shorthands of the canonical ones.
This allows us to focus on the canonical ones later.

Author: I made a start on this in de4e00a42764c9739c2f18ce1559d80c8607f063, but we probably need to do more.

Reviewer: I'll keep this conversation open, but we can do that in the next round.

Author: @cabo See 48de8fcda69084c2ea850805af8d5041492b1943 for an early statement along the lines you suggested.

> | JSONPath            | Description                                                                                                             |

Reviewer: This should be sorted into query starts ('root identifiers'), query steps ('segments', sorted into the canonical ones and their shorthands), and selectors.
We need to make sure these are not mixed up in the minds of the readers.

Author: d4c8c39da0f1dc4c69fd1f9d5a6cb297a2e6c3e2 puts the identifiers first, then segments, then some selectors, then the shorthands. (Moving the shorthands up wouldn't seem to be very readable since they need to build on selectors.)

Reviewer: Right.  Might be useful to have some section headers (we can do this in the next round).

Author: @cabo See https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/268 for a stab at improving the structure of the overview.

which is applied to the result of the previous root identifier or __APPENDER__ and provides
input to the next __APPENDER__.
These results and inputs take the form of a *nodelist*, i.e., a
sequence of zero or more nodes.

The nodelist resulting from the root identifier contains a single node,
the argument.
The nodelist resulting from the last __APPENDER__ is presented as the
result of the query; depending on the specific API, it might be
presented as an array of the JSON values at the nodes, an array of
Normalized Paths referencing the nodes, or both — or some other
representation as desired by the implementation.
Note that the API must be capable of presenting an empty nodelist as
the result of the query.

An __APPENDER__ performs its function on each of the nodes in its input
nodelist, during such a function execution, such a node is referred to
as the "current node".  Each of these function executions produces a
nodelist, which are then concatenated to produce
the result of the __APPENDER__. A node may be selected more than once and
appears that number of times in the nodelist. Duplicate nodes are not removed.

> A syntactically valid __APPENDER__ MUST NOT produce errors.

Reviewer: ```suggestion
A syntactically valid __APPENDER__ MUST NOT produce errors when executing the query.
```

... is a bit more explicit.

Author: Adopted in ef09a71faad2e8263b38a61c8360feea1564e343.

This is the concatenation of three lists, two of length one containing
`0`, `1`, respectively, and one of length zero.

As a consequence of this approach, if any of the __APPENDERS__ produces an empty nodelist,
then the whole query produces an empty nodelist.

In what follows, the semantics of each __APPENDER__ are defined for each type
of node.

## Root Identifier

### Syntax
{: unnumbered}

> Every valid JSONPath query MUST begin with the root identifier `$`.

Reviewer: What does the word "valid" mean here?
Do we also describe "invalid JSONPath queries"?

This sentence has the disadvantage that it makes @ queries into something different, which then require another word for $ queries and @ queries collectively.

Author: e96a561ef79ac2457c3d44f9bdb0fca77112ca2b deletes the word "valid".

I need to think more about $ and @ terminology.

Reviewer: I'll keep this conversation open, but we can do this in the next round.

| :---: | ------ | :----------: | ------- |
| `$.o['j j']['k.k']`   | `3` | `$['o']['j j']['k.k']`      | Named value in nested object      |
| `$.o["j j"]["k.k"]`   | `3` | `$['o']['j j']['k.k']`      | Named value in nested object      |
| `$["'"]["@"]` | `2` | `$['\'']['@']` | Unusual member names
| `$.j`   | `{"k": 3}` | `$['j']`      | Named value of an object      |
| `$.j.k` | `3`        | `$['j']['k']` | Named value in nested object  |
{: title="Name selector examples"}

### Wildcard Selector {#wildcard}

#### Syntax
{: unnumbered}

The wildcard selector consists of an asterisk.

~~~~ abnf
wildcard = "*"
~~~~

#### Semantics
{: unnumbered}

> A `wildcard` selector selects the nodes of all children of an object or array.

Reviewer: Did we define "children of an object"?
(These are the values, not the members.)

Author: Yes we did. The terminology section says this:

> Children (of a node):
>
>    If the node is an array, each of its elements, or if the node is an object, each of its member values (but not its member names). If the node is neither an array nor an object, it has no children.

Reviewer: Thanks for looking that up for me 😊

Queries:

The following examples show the `wildcard` selector in use by a child __APPENDER__.

| Query | Result | Result Paths | Comment |
| :---: | ------ | :----------: | ------- |
| `$[*]`   | `{"j": 1, "k": 2}` <br> `[5, 3]` | `$['o']` <br> `$['a']` | Object values      |
| `$.o[*]` | `1` <br> `2` | `$['o']['j']` <br> `$['o']['k']` | Object values      |
| `$.o[*]` | `2` <br> `1` | `$['o']['k']` <br> `$['o']['j']` | Alternative result |
| `$.a[*]` | `5` <br> `3` | `$['a'][0]` <br> `$['a'][1]`     | Array members      |
{: title="Wildcard selector examples"}

### Index selector {#index-selector}

#### Syntax {#syntax-index}
{: unnumbered}

An index selector `<index>` matches at most one array element value.

~~~~ abnf
index-selector  = int                             ; decimal integer

int             = ["-"] ( "0" / (DIGIT1 *DIGIT) ) ; -  optional
DIGIT1          = %x31-39                         ; 1-9 non-zero digit
~~~~

> Applying the numerical `index-selector` is required to select the corresponding

Reviewer: ```suggestion
Applying the numerical `index-selector` selects the corresponding
```

element. JSONPath allows it to be negative (see {{index-semantics}}).

Notes:
1. An `index-selector` is an integer (in base 10, as in JSON numbers).
2. As in JSON numbers, the syntax does not allow octal-like integers with leading zeros such as `01` or `-01`.

#### Semantics {#index-semantics}
{: unnumbered}

> The `index-selector` applied to an array selects an array element using a zero-based index.

Reviewer: Should we say what happens when it is not applied to an array?
(Nothing is selected.)

Author: Er, that paragraph ends with:

> Nothing is selected from a value that is not an array.

Reviewer: Sounds good :-)

> Array indexing is a way of selecting a particular element of an array using

Reviewer: Can we say this is the same as with an index selector?

Author: I think is simpler fix is to delete the index selector stuff from this section. See 977965ad795158e802bfbf82c0f1a62c7b0b9a2f.

Reviewer: Thanks.  I note that there is still some repetition in the semantics section, but we can fix that later.

Shorthand notations are also provided for when there is a single
wildcard or name selector.

~~~~ abnf
child-APPENDER            = (child-longhand /
                             dot-wildcard-shorthand /
                             dot-member-name-shorthand)

child-longhand            = "[" S selector 1*(S "," S selector) S "]"

dot-wildcard-shorthand    = "." wildcard

dot-member-name-shorthand = "." dot-member-name
dot-member-name           = name-first *name-char
name-first                = ALPHA /
                            "_"   /            ; _
                            %x80-10FFFF        ; any non-ASCII Unicode character
name-char                 = DIGIT / name-first

DIGIT                     =  %x30-39              ; 0-9
ALPHA                     =  %x41-5A / %x61-7A    ; A-Z / a-z
~~~~

The `dot-wildcard-shorthand` is shorthand for `[*]`.

A `dot-member-name-shorthand` of the form `.<member-name>` is shorthand for `['<member-name>']`.

Member names containing characters other than allowed by
`dot-member-name-shorthand` — such as space (U+0020), minus (U+002D), dot (U+002E)
or escaped characters which appear in the [name selector semantics section](#name-semantics) —

> MUST NOT be used with the `dot-member-name-shorthand`.

Reviewer: I don't understand this sentence.  How can the names be used if you can't even notate them?
This should simply say that the set of member names that can be used is limited by the syntax of dot-member-name-shorthands.

Author: I simplified this in 397f2befc51d47caf6403dccfc6c720468081ad1.

> (In such cases, the `name-selector` syntax MUST be used instead.)

Reviewer: This is a weird MUST, as it is a statement of fact ("if the key at hand does not fit the door, the key that does fit MUST be used"???).

Author: 397f2befc51d47caf6403dccfc6c720468081ad1 deleted this.

appear in the list.
Note that any node matched by more than one selector is kept
as many times in the nodelist.

### Descendant __APPENDER__

#### Syntax
{: unnumbered}

The descendant __APPENDER__ consists of a double dot `..`
followed by a child __APPENDER__ (`descendant-APPENDER`).

Shortand notations are also provided that correspond to the shorthand forms of the child __APPENDER__.

~~~~ abnf
descendant-APPENDER              = (descendant-child /
                                    descendant-wildcard-shorthand /
                                    descendant-member-name-shorthand)
descendant-child                 = ".." child-APPENDER

descendant-wildcard-shorthand    = ".." wildcard
descendant-member-name-shorthand = ".." dot-member-name
~~~~

The `descendant-wildcard-shorthand` is shorthand for `..[*]`.

A `descendant-member-name-shorthand` of the form `..<member-name>` is shorthand for `..['<member-name>']`.

Note that `..` on its own is not a valid __APPENDER__.

#### Semantics
{: unnumbered}

<!-- The following does not address https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/252 -->

> A descendant __APPENDER__ selects zero or more descendants of the input value.

Reviewer: I don't think a segment "selects".  It applies its selector to each of the self-or-descendant axis nodes (we need more words).  The sequence of application is partially deterministic.  Outputs of the selectors are concatenated in the sequence of application.

Author: I avoided saying "selects" in d5286be84535a41c0aa8e789b1c2d58c5cec5d86.

The syntax and semantics of each segment are defined below.

### Child Segment

#### Syntax
{: unnumbered}

The child segment consists of a non-empty, comma-delimited
sequence of selectors enclosed in square brackets.

Shorthand notations are also provided for when there is a single
wildcard or name selector.

~~~~ abnf
child-segment             = (child-longhand /
                             dot-wildcard-shorthand /
                             dot-member-name-shorthand)

child-longhand            = "[" S selector 1*(S "," S selector) S "]"

dot-wildcard-shorthand    = "." wildcard

dot-member-name-shorthand = "." dot-member-name
dot-member-name           = name-first *name-char
name-first                = ALPHA /
                            "_"   /            ; _
                            %x80-10FFFF        ; any non-ASCII Unicode character
name-char                 = DIGIT / name-first

DIGIT                     =  %x30-39              ; 0-9
ALPHA                     =  %x41-5A / %x61-7A    ; A-Z / a-z
~~~~

The `dot-wildcard-shorthand` is shorthand for `[*]`.

A `dot-member-name-shorthand` of the form `.<member-name>` is shorthand for `['<member-name>']`, but

> can only be used with members names containing certain characters.

Reviewer: ```suggestion
can only be used with member names that are composed of certain characters.
```
